File: packages/sdk-php/src/MelayaException.php
<?php

declare(strict_types=1);

namespace Melaya;

use RuntimeException;

/**
 * Thrown for non-2xx HTTP responses or when the API returns `ok: false`.
 *
 * NOTE: We do NOT redeclare $code or $message here — Exception already owns those.
 * Use $errorCode for the API error code string and $status for HTTP status.
 */
class MelayaException extends RuntimeException
{
    /** HTTP status code (0 for transport/cURL errors). */
    public readonly int $status;

    /** The `error` field from the API response envelope (e.g. "COLD_CACHE"). */
    public readonly ?string $errorCode;

    /** The raw parsed response body, if available. */
    public readonly mixed $body;

    public function __construct(
        string $message,
        int $status = 0,
        ?string $errorCode = null,
        mixed $body = null,
    ) {
        parent::__construct($message);
        $this->status    = $status;
        $this->errorCode = $errorCode;
        $this->body      = $body;
    }
}

File: packages/sdk-java/src/main/java/org/melaya/MelayaException.java
package org.melaya;

/**
 * Thrown for non-2xx HTTP responses or when the API envelope returns {@code ok: false}.
 */
public class MelayaException extends RuntimeException {

    private final int status;
    private final String code;
    private final Object body;

    public MelayaException(String message, int status, String code, Object body) {
        super(message);
        this.status = status;
        this.code = code;
        this.body = body;
    }

    /** HTTP status code (e.g. 400, 401, 429). */
    public int getStatus() {
        return status;
    }

    /** Machine-readable error code from the API envelope (may be {@code null}). */
    public String getCode() {
        return code;
    }

    /** Raw parsed response body (may be {@code null}). */
    public Object getBody() {
        return body;
    }

    @Override
    public String toString() {
        return "MelayaException{status=" + status + ", code=" + code + ", message=" + getMessage() + "}";
    }
}
